### ISSUE TITLE
Writing a MonoGame Host

### ISSUE THREAD

**Issue by Cyral** [yuvaltz/Granular#9]
Let me just say, this project is amazing!

I've been writing a WPF style UI framework for a few weeks to be used for a MonoGame/XNA game my team is developing. (Currently there are no good XAML based UI libraries, only WinForms style.) Writing our own framework for (especially the arrange/layout process) is taking significant time and I think using Granular may be a better option.

It seems creating a host is well supported and it should be somewhat easy to write a renderer for MonoGame. (Although MonoGame and XNA don't have support for any 2D primitive drawing functions, I can make those)

What kind of work would it take to create a MonoGame host? (If you are unfamiliar with MonoGame, it has the standard game loop, sprite drawing functions, etc.) The WPF host seems to convert to WPF controls, and the Web host into HTML, so I'm not quite sure where to get started in plugging in rendering code.

**Comment by yuvaltz** [comment]
Thank you! I really appreciate it.

It shouldn't be too hard to implement a new host.
It starts from the end application's AssemblyInfo.cs:
`[assembly: ApplicationHost(typeof(WpfApplicationHost))]`
The IApplicationHost provides factories for an IPresentationSource and a few others.
IPresentationSource is representing a rendering area (a native window, or a page in the browser), it gets a root element (which is usually a "Window"), and injects an IRenderElementFactory to get its rendering element. Also it notifies the input devices about raw events (mouse and keyboard).
That's the general idea.  
You can leave some of the interfaces unimplemented, start with an application with an empty window, and see what's important to implement.
also, it may help you to look at / compare the two Hosts implementations.  

If you have any questions feel free :)

**Comment by Cyral** [comment]
Quick question, is the presentation part of Granular compatible with Mono? (Doesn't use any Windows only features?)

**Comment by yuvaltz** [comment]
Yes, the presentation part only defines those interfaces, it doesn't assume anything about the implementation (in the browser there aren't any Windows features anyway).
